services/dashboard/services/restapi.js

```diff
@@ -0,0 +1,167 @@
+            $timeout.cancel(timeoutPromise);
+          });
+
+          var fn = function() {
+            $http.get(restapiV1Root + path)
+              .then(function(response) {
+                if (!shouldCancel) {
+                  shouldCancel = !onData || onData(response.data);
+                }
+              }, function(response) {
+              })
+              .finally(function() {
+                if (!shouldCancel) {
+                  timeoutPromise = $timeout(fn, interval || conf.restapiQueryInterval);
+                }
+              });
+          };
+          fn();
+        },
+
+        get: function(path, onData) {
+          return $http.get(restapiV1Root + path);
+        },
+
+        /** Get data from server periodically until an user cancellation or scope exit. */
+        repeatUntil: function(url, scope, onData) {
+          // TODO: Once `subscribe` is turned to websocket push model, there is no need to have this method
+          this.subscribe(url, scope,
+            function(data) {
+              return !onData || onData(data);
+            });
+        },
+
+        /** Kill a running application */
+        killApp: function(appId) {
+          var url = restapiV1Root + '/appmaster/' + appId;
+          return $http.delete(url);
+        },
+
+        /** Restart a running application and return a promise */
+        restartAppAsync: function(appId) {
+          var url = restapiV1Root + '/appmaster/' + appId + '/restart';
+          return $http.post(url);
+        },
+
+        /** Return the config link of an application */
+        appConfigLink: function(appId) {
+          return restapiV1Root + '/appmaster/' + appId + '/config';
+        },
+
+        /** Return the config link of a worker */
+        workerConfigLink: function(workerId) {
+          return restapiV1Root + '/worker/' + workerId + '/config';
+        },
+
+        /** Return the config link of the master */
+        masterConfigLink: function() {
+          return restapiV1Root + '/master/config';
+        },
+
+        /** Submit an user defined application with user configuration */
+        submitUserApp: function(files, formFormNames, args, onComplete) {
+          var params = args ? '?args=' + encodeURIComponent(args) : '';
+          var upload = Upload.upload({
+            url: restapiV1Root + '/master/submitapp' + params,
+            method: 'POST',
+            file: files,
+            fileFormDataName: formFormNames
+          });
+
+          upload.then(function(response) {
+            if (onComplete) {
+              var data = response.data;
+              onComplete({success: data && data.success});
+            }
+          }, function() {
+            if (onComplete) {
+              onComplete({success: false});
+            }
+          });
+        },
+
+        /** Replace a dag processor at runtime */
+        replaceDagProcessor: function(appId, oldProcessorId, newProcessorDescription, onComplete) {
```

Comment:
Would be better if this app specific methods could be moved out of restapi - can probably be done in the scalajs phase.